### ISSUE TITLE
DQM instability in 12434.0, pixelLessStep_quickAssociatorByHits

### ISSUE THREAD

**Issue by jpata** [cms-sw/cmssw#37970]
In recent PRs it seems the workflow 12434.0 has DQM changes in Tracking/TrackParameters/generalTracks/SeedMon/pixelLessStep/TrackBuilding, pixelLessStep_quickAssociatorByHits

https://github.com/cms-sw/cmsdist/pull/7862
https://github.com/cms-sw/cmssw/pull/37918
https://github.com/cms-sw/cmssw/pull/37952

The change is not always in the same direction:
![image](https://user-images.githubusercontent.com/69717/168759645-2fa9f39e-8b65-45eb-8266-31357da01dda.png)
![image](https://user-images.githubusercontent.com/69717/168759665-1d0fa838-c337-4f6c-968b-c680918946b0.png)

@cms-sw/tracking-pog-l2

**Comment by jpata** [comment]
assign reconstruction

**Comment by jpata** [comment]
As I noted [here](https://github.com/cms-sw/cmsdist/pull/7862#issuecomment-1128500285), it does not seem immediately related to TF giving different results on different CPUs.
```
37918 baseline: 2022-05-16 14:22:17.657780: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  SSE4.1 SSE4.2 AVX AVX2 FMA
37918 PR: 2022-05-16 14:49:16.906255: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  SSE4.1 SSE4.2 AVX AVX2 FMA
```

**Comment by makortel** [comment]
(speaking from distant past experience) Looking at https://github.com/cms-sw/cmssw/pull/37918#issuecomment-1127941919, there are differences in folder [`Tracking/TrackParameters/generalTracks/SeedMon/pixelLessStep`](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_4_X_2022-05-16-1100+9966c6/50311/12434.0_TTbar_14TeV+2023+TTbar_14TeV_TuneCP5_GenSim+Digi+RecoNano+HARVESTNano+ALCA/Tracking_TrackParameters_generalTracks_SeedMon_pixelLessStep_TrackBuilding.html) like
![image](https://user-images.githubusercontent.com/33993/168824014-ad75a2a1-d00d-4c82-b869-2821ca78995f.png)

These plots are "purely DQM", i.e. tracks (well, `TrackCandidate`s) are not matched to MC. This hints that the difference comes  in some way from reconstruction. Could there be e.g. uninitialized variable somewhere where the (mkfit) track candidates are fed to the DNN? @leonardogiannini @slava77

**Comment by slava77** [comment]
is it clear that the differences started showing up after #37878 ?

**Comment by slava77** [comment]
> Could there be e.g. uninitialized variable somewhere where the (mkfit) track candidates are fed to the DNN? @leonardogiannini @slava77

there is indeed an execution path that could lead to uninitialized values (with an assumption that a default-constructed `tf::Tensor` is not initialized, a trackCandidate that fails a PCA would run evaluation over uninitialized values).
@leonardogiannini is investigating

**Comment by leonardogiannini** [comment]
this should indeed be due to https://github.com/cms-sw/cmssw/blob/master/RecoTracker/MkFit/plugins/MkFitOutputConverter.cc#L637 which leaves the input undetermined in some cases. 
I found 1/3510 tracks (100 evts) passed to cand DNN where this happens. in TTbar with PU this is happens in about 0.1% of tracks passed to cand DNN.

A quick fix would be adding the initialization to 0, e.g. :

       if (!(tsAtClosestApproachTrackCand.isValid())) {
         edm::LogVerbatim("TrackBuilding") << "TrajectoryStateClosestToBeamLine not valid";
         for (auto i=0; i<29; i++) {
            input1.matrix<float>()(nt, i) = 0;
         }
         input2.matrix<float>()(nt, 0) = algo_;
         continue;
       }

I can open the PR. I imagine it needs a backport as well.

**Comment by slava77** [comment]
> A quick fix would be adding the initialization to 0, e.g. :

does a 0 input return an output dnn value that would pass or fail the cuts?
My guess is that it's better to fail the candidate.
So, there may be a need to override the `dnn(0)` .

**Comment by leonardogiannini** [comment]
Unfortunately, the all 0 tensor passes the selection (though not significant).
The other possibility is to store the indices "itrack" (in a std::vector?) where it happens and then fail the cand making the output -1.

**Comment by jpata** [comment]
type tracking

**Comment by jpata** [comment]
I understand this would be something urgent to address before the final release of 12_4_0 (closed to physics changes, but bugfixes should be fine). Is this something you can address in the next week or so, @leonardogiannini?

**Comment by mmusich** [comment]
>  The other possibility is to store the indices "itrack" (in a std::vector?) where it happens and then fail the cand making the output -1.

Would something like this work?

```diff
diff --git a/RecoTracker/MkFit/plugins/MkFitOutputConverter.cc b/RecoTracker/MkFit/plugins/MkFitOutputConverter.cc
index 16fdfad9ec1..d27575aaad2 100644
--- a/RecoTracker/MkFit/plugins/MkFitOutputConverter.cc
+++ b/RecoTracker/MkFit/plugins/MkFitOutputConverter.cc
@@ -616,6 +616,8 @@ std::vector<float> MkFitOutputConverter::computeDNNs(TrackCandidateCollection co
   tensorflow::Tensor input1(tensorflow::DT_FLOAT, {bsize_, 29});
   tensorflow::Tensor input2(tensorflow::DT_FLOAT, {bsize_, 1});
 
+  std::vector<int> toFail;
+
   for (auto nb = 0; nb < nbatches + 1; nb++) {
     for (auto nt = 0; nt < bsize_; nt++) {
       int itrack = nt + bsize_ * nb;
@@ -634,6 +636,7 @@ std::vector<float> MkFitOutputConverter::computeDNNs(TrackCandidateCollection co
 
       if (!(tsAtClosestApproachTrackCand.isValid())) {
         edm::LogVerbatim("TrackBuilding") << "TrajectoryStateClosestToBeamLine not valid";
+        toFail.push_back(itrack);
         continue;
       }
 
@@ -720,7 +723,12 @@ std::vector<float> MkFitOutputConverter::computeDNNs(TrackCandidateCollection co
         continue;
 
       float out0 = 2.0 * outputs[0].matrix<float>()(nt, 0) - 1.0;
-      output[itrack] = out0;
+
+      if (std::find(toFail.begin(), toFail.end(), itrack) != toFail.end()) {
+        output[itrack] = -1.f;
+      } else {
+        output[itrack] = out0;
+      }
     }
   }
```

**Comment by leonardogiannini** [comment]
yes, that's it more or less. I am testing almost the same. You or I can open the PR. Let me know

**Comment by mmusich** [comment]
> yes, that's it more or less. I am testing almost the same. You or I can open the PR. Let me know

go ahead!

**Comment by perrotta** [comment]
Looks like having been actually fixed by #38004, see also https://github.com/cms-sw/cmssw/pull/37954#issuecomment-1132867789

**Comment by jpata** [comment]
+reconstruction
- solved by https://github.com/cms-sw/cmssw/pull/38004
